### PR TITLE
[Enhancement] Support order by for non primary key table

### PR DESCRIPTION
**Pull request by sevev**
After this pr, for primary key table and duplicate table, sort key column has no matter with key column, and we can specify sorting columns using `order by` when creating a table, for example:
```
CREATE TABLE `t1` (
  `pk` bigint(20) NOT NULL COMMENT "",
  `col1` bigint(20) NOT NULL COMMENT "",
  `col2` int(11) NULL COMMENT ""
) ENGINE=OLAP
PRIMARY KEY(`pk`) // DUPLICATE KEY(`pk`)
DISTRIBUTED BY HASH(`pk`) BUCKETS 1
ORDER BY (col1)
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "true",
"compression" = "LZ4"
);
```

For another type table(AGGREGATE TABLE and UNIQUE TABLE so far), sorting columns still need to include all key columns but can have a different order than the key columns, for example:
```
CREATE TABLE `t1` (
  `k1` bigint(20) NOT NULL COMMENT "",
  `k2` bigint(20) NOT NULL COMMENT "",
  `col1` bigint(20) NOT NULL COMMENT "",
  `col2` int(11) NULL COMMENT ""
) ENGINE=OLAP
AGGGEGATE KEY(`k1, k2`) 
DISTRIBUTED BY HASH(`pk`) BUCKETS 1
ORDER BY(`k2`,`k1`)
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT",
"enable_persistent_index" = "true",
"compression" = "LZ4"
);
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by imay** [comment]
@sevev 
My question is what the behavior is when there is no explicit `ORDER BY()` clause in the create table statement?
I suppose it should be the same as it does now.

**Comment by sevev** [comment]
> > @sevev My question is what the behavior is when there is no explicit `ORDER BY()` clause in the create table statement? I suppose it should be the same as it does now. And another question is if users create a table with different order by columns in new versions, and what will happen if they downgrade the cluster to a lower version in case of some issues?
> 
> 1. If we don't specify sort key columns in create table statement, the sort key columns are the same with key columns, consistent with current behavior.
> 2. Compatibility issue does exist here. If users create a table with different order by columns in new versions, and downgrade the cluster to a lower version, they need to rewrite the data into the new table which sort key columns is same with key columns, otherwise the query may return error result in old version.

**Comment by imay** [comment]
> Compatibility issue does exist here. If users create a table with different order by columns in new versions, and downgrade the cluster to a lower version, they need to rewrite the data into the new table which sort key columns is same with key columns, otherwise the query may return error result in lower version.

I think we need to find a way to avoid this online issue happening, otherwise, it will introduce some issues when users downgrade which will be very weird to users.
At least we can set a variable or something to disable this feature in 3.3 if this is introduced into it. And we need to notice the dangers in the release notes. Users can enable it explicitly by setting the global variable. IMO, nearly no one will enable it when upgrading, so it may be safe in most scenarios.
Of course, it will be best if you can find a way to avoid incompatibility in any situation.
So before the compatibility issue has been fixed, this PR can not be merged.

**Comment by sevev** [comment]
> > Compatibility issue does exist here. If users create a table with different order by columns in new versions, and downgrade the cluster to a lower version, they need to rewrite the data into the new table which sort key columns is same with key columns, otherwise the query may return error result in lower version.
> 
> I think we need to find a way to avoid this online issue happening, otherwise, it will introduce some issues when users downgrade which will be very weird to users. At least we can set a variable or something to disable this feature in 3.3 if this is introduced into it. And we need to notice the dangers in the release notes. Users can enable it explicitly by setting the global variable. IMO, nearly no one will enable it when upgrading, so it may be safe in most scenarios. Of course, it will be best if you can find a way to avoid incompatibility in any situation. So before the compatibility issue has been fixed, this PR can not be merged.

I think supporting backward compatibility to the latest version is sufficient.  For example, if we upgrade to version 3.3, we support rolling back to versions 3.1.5 and later, but we do not support rolling back to versions prior to 3.1.5. We cannot make the released versions backward compatible because the released versions are immutable.

This needs to be clearly documented in the release notes and I think this is acceptable. Do you think this is feasible? @imay

**Comment by imay** [comment]
> I think supporting backward compatibility to the latest version is sufficient. For example, if we upgrade to version 3.3, we support rolling back to versions 3.1.5 and later, but we do not support rolling back to versions prior to 3.1.5. We cannot make the released versions backward compatible because the released versions are immutable.

Totally agree.

> This needs to be clearly documented in the release notes and I think this is acceptable. Do you think this is feasible? @imay

Look good to me.

**Review by decster — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/KeysDesc.java [hunk 1]
```diff
@@ -241,13 +261,29 @@ public void checkColumnDefs(List<ColumnDef> cols, List<Integer> sortKeyIdxes) {
...
+                }
+            } else if (type == KeysType.DUP_KEYS) {
+                // sort key column of duplicate table has no limitation
+            } else if (type == KeysType.AGG_KEYS || type == KeysType.UNIQUE_KEYS) {
+                // sort key column of AGG and UNIQUE table must include all key columns and cannot have any columns other than
+                // the key columns
+                boolean res = new HashSet<>(keyColIdxes).equals(new HashSet<>(sortKeyIdxes));
+                if (!res) {
+                    throw new SemanticException("The sort columns of " + type.toSql() + " table must include all key " +
+                                                "columns and cannot have any columns other than key columns.");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/analysis/KeysDesc.java:283]:
```suggestion
                    throw new SemanticException("The sort columns of " + type.toSql() + " table must be same with key columns");
```

File: fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java [hunk 2]
```diff
@@ -737,11 +749,38 @@ private void processReorderColumnOfPrimaryKey(ReorderColumnsClause alterClause,
...
             }
+        } else if (olapTable.getKeysType() == KeysType.AGG_KEYS || olapTable.getKeysType() == KeysType.UNIQUE_KEYS) {
+            // sort key column must include all key columns and can not have any other columns
+            boolean res = new HashSet<>(keyColumnIdxes).equals(new HashSet<>(sortKeyIdxes));
+            if (!res) {
+                throw new DdlException("The sort columns of " + olapTable.getKeysType().toSql() + " table must include all key " +
+                                       "columns and cannot have any other columns");
+            }
+        } else {
+            throw new DdlException("Table type:" + olapTable.getKeysType().toSql() + " does not support sort key column");
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/alter/SchemaChangeHandler.java:783]: the message should be same with `KeysDesc`

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 3]
```diff
@@ -165,23 +165,22 @@ public static void analyze(CreateTableStmt statement, ConnectContext context) {
...
         List<Integer> sortKeyIdxes = Lists.newArrayList();
         if (statement.getSortKeys() != null) {
-            if (keysDesc == null || keysDesc.getKeysType() != KeysType.PRIMARY_KEYS) {
-                NodePosition keysPos = NodePosition.ZERO;
-                if (keysDesc != null) {
-                    keysPos = keysDesc.getPos();
+            List<String> columnNames = 
+                    statement.getColumnDefs().stream().map(ColumnDef::getName).collect(Collectors.toList());
+            for (String column : statement.getSortKeys()) {
+                int idx = columnNames.indexOf(column);
```
[the remark this answers is not in the document]
Comment (sevev) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:171]: > case insensitive？

Column names are not case-sensitive, and all column names in the `CreateTableStatement` have been converted to lowercase.

**Review by wyb — DISMISSED** [review]

**Review by wyb — APPROVED** [review]

**Review by decster — APPROVED** [review]

File: be/src/storage/memtable.cpp [hunk 4]
```diff
@@ -387,10 +389,22 @@ void MemTable::_aggregate(bool is_final) {
...
 
-void MemTable::_sort(bool is_final, bool by_sort_key) {
+Status MemTable::_sort(bool is_final, bool by_sort_key) {
     SmallPermutation perm = create_small_permutation(static_cast<uint32_t>(_chunk->num_rows()));
     std::swap(perm, _permutations);
-    _sort_column_inc(by_sort_key);
+
+    // sort key column has some limitation right now:
+    // 1. DUPLICATE TABLE and PRIMARY TABLE: no limitation
+    // 2. AGGREGATE TABLE and UNIQUE TABLE: sort key columns must inclue all key columns and can not
```
Comment (liuyehcf) [inline be/src/storage/memtable.cpp:398]: inclue -> include

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**